authority/provisioner/webhook.go

```diff
@@ -0,0 +1,203 @@
+package provisioner
+
+import (
+	"bytes"
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"text/template"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/smallstep/certificates/templates"
+	"go.step.sm/crypto/sshutil"
+	"go.step.sm/crypto/x509util"
+	"golang.org/x/crypto/ssh"
+)
+
+type Webhook struct {
+	ID                   string `json:"id"`
+	Name                 string `json:"name"`
+	URL                  string `json:"url"`
+	Kind                 string `json:"kind"`
+	DisableTLSClientAuth bool   `json:"disableTLSClientAuth,omitempty"`
+	SigningSecret        string `json:"-"`
+	BearerToken          string `json:"-"`
+	BasicAuth            struct {
+		Username string
+		Password string
+	} `json:"-"`
+}
+
+type sshRequest struct {
+	Key        []byte
+	Type       string
+	KeyID      string
+	Principals []string
+}
+
+// ssh.Certificate but the Key and SignatureKey are marshaled
+type sshCert struct {
+	Nonce           []byte          `json:"nonce"`
+	Key             []byte          `json:"key"`
+	Serial          uint64          `json:"serial"`
+	CertType        uint32          `json:"certType"`
+	KeyID           string          `json:"keyID"`
+	ValidPrincipals []string        `json:"validPrincipals"`
+	ValidAfter      uint64          `json:"validAfter"`
+	ValidBefore     uint64          `json:"validBefore"`
+	Permissions     ssh.Permissions `json:"permissions"`
+	Reserved        []byte          `json:"reserved"`
+	SignatureKey    []byte          `json:"signatureKey"`
+	Signature       *ssh.Signature  `json:"Signature"`
```

Comment:
This won't generally exists, but it should be `signature` and `omitempty` as it will be nil.
```suggestion
	Signature       *ssh.Signature  `json:"signature,omitempty"`
```